### PR TITLE
bpo-39305: Update nntplib to merge nntplib.NNTP and nntplib._NNTPBase

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue39305

### CODE REVIEW

**Comment by corona10** [comment]
@vstinner Ready to review!

**Comment by corona10** [comment]
@vstinner gentle ping

**Comment by corona10** [comment]
@ericvsmith Would you mind to take a look at this PR?

**Comment by corona10** [comment]
Can you please take a look also? @warsaw :)

File: Lib/nntplib.py [hunk 1]
```diff
@@ -96,7 +96,7 @@
 class NNTPError(Exception):
     """Base class for all nntplib exceptions"""
     def __init__(self, *args):
-        Exception.__init__(self, *args)
+        super().__init__(*args)
         try:
```
Comment (ericvsmith) [inline Lib/nntplib.py]: This change seems unrelated. Is it required?

File: Lib/nntplib.py [hunk 2]
```diff
@@ -325,7 +330,21 @@ def __init__(self, file, host,
...
+            self._base_init(readermode)
+            if user or usenetrc:
+                self.login(user, password, usenetrc)
+        except:
+            if self.file:
+                self.file.close()
+            self.sock.close()
+            raise
+
+    def _base_init(self, readermode):
```
Comment (ericvsmith) [inline Lib/nntplib.py:347]: Does `_base_init` need to be public? I notice it's used in the test, but would it be expected to be used by non-test code?

**Review by ericvsmith — COMMENTED** [review]
See my comments. I think it generally looks okay.

File: Lib/nntplib.py (hunk 1 above)
Comment (corona10) [reply Lib/nntplib.py]: @ericvsmith Yes this change is unrelated, should I create for the separate PR?

File: Lib/nntplib.py (hunk 2 above)
Comment (corona10) [reply Lib/nntplib.py:347]: I agree that the changed structure is not that good.
Do you have any suggestions to be better in the abstract?

File: Lib/nntplib.py (hunk 1 above)
Comment (ericvsmith) [reply Lib/nntplib.py]: I'd put it in a separate PR. But is this a common pattern in the stdlib? I looked through a few libraries and couldn't find any instances of an exception that takes any parameters, so I'm not sure what best practice is here. I guess it can't hurt.

File: Lib/nntplib.py (hunk 2 above)
Comment (ericvsmith) [reply Lib/nntplib.py:347]: Well, normally I'd say add an intermediate base class with a different constructor! :)

I don't think the code is so bad, but I'm questioning if code other than test code needs the same pattern, and if so, we should make this public.

Comment (corona10) [reply Lib/nntplib.py:347]: @ericvsmith 
My first intension was removing the intermediate base class :)
See: https://github.com/python/cpython/pull/17939#pullrequestreview-341290152

Unfortunately, the test code is using the base class.
So I choose this usage and _NNTPBase was not the public class.

File: Lib/nntplib.py (hunk 1 above)
Comment (corona10) [reply Lib/nntplib.py]: I observed this kind of simple refactoring works was done by other huge work.

Comment (ericvsmith) [reply Lib/nntplib.py]: My concern here is not so much including this in this PR (I'd move it to another PR, but I won't object to leaving it here).

My concern is: is an exception calling `super()`, such that it can be used with cooperative multiple inheritance, a common thing in the stdlib? Or is this the only place in the stdlib that it's done? If this is the only place it will be used in the stdlib, we should either change every usage in the the stdlib to use `super()`, or not make this change at all.

File: Lib/nntplib.py (hunk 2 above)
Comment (ericvsmith) [reply Lib/nntplib.py:347]: So the root of the problem is that the test code wants to pass in (or set) the `file` parameter when initializing the `NNTP` object. If that's something that we generally we don't want to support, then I think using the private `_base_init` is okay for the test code. It's a little odd that `NNTP.__init__` is never being called, and `self.port` is never being initialized, but I guess that's okay for the unsupported test case code.

So, after thinking about it: I'd make some comment in `_base_init` about why it exists (only to support testing where `file` is passed in), and not change any of the actual code here.

**Review by ericvsmith — APPROVED** [review]
Thanks! Looks good to me.

**Review by corona10 — COMMENTED** [review]
@ericvsmith Thank you Eric, I will merge it after the CI is passed :)
